### ISSUE TITLE
Silent sign-in request was sent but no user is signed in

### ISSUE THREAD

**Issue by XavierDupessey** [AzureAD/azure-activedirectory-library-for-js#456]
Hello,

I get an error with adal and adal-angular on Chrome with this use case:

1. A user attemps to access the app securized with adal on the root URL. A ui router rule redirects him to the real home page (/tickets/) which needs authentication and has 2 resolves invoking protected HTTP requests:

    ```
    $urlRouterProvider.when('', '/tickets/')

    $stateProvider.state('tickets', {
        'abstract': true,
        'template': '<ui-view />',
        'url': '/tickets',
    })

    $stateProvider.state('tickets.index', {
        'component': 'hasccTicketsIndex',
        'requireADLogin': true,
        'resolve': {
            'allOpenTickets': allOpenTicketsResolve,
            'userTickets': userTicketsResolve,
        },
        'url': '/',
    })
    ```

2. The user is redirected to login.microsoftonline.com (as expected) and logs in. It is redirected to the app page and everything works fine.

3. The user stops using the app and closes its Chrome browser. He waits more than 1 hour (the token expiration is 1 hour long).

4. The user starts the app, on the root URL. Nothing happends and an error appears in the logs:

> AADSTS50058: A silent sign-in request was sent but no user is signed in. The cookies used to represent the user's session were not sent in the request to Azure AD. This can happen if the user is using Internet Explorer or Edge, and the web app sending the silent sign-in request is in different IE security zone than the Azure AD endpoint (login.microsoftonline.com).
> Trace ID: aa7c06e6-a433-4e83-be32-31aae1c17a07
> Correlation ID: b50d6b22-320e-4a0a-872c-31c6b3450974
> Timestamp: 2016-12-07 08:12:51Z

Environment:
- Chrome v54.0.2840.99 m
- Angular v1.5.8
- Adal-angular and AdalJs v1.0.12 (I am going to test with the v1.0.13 later today)

Do you know why I get this error and how to fix this problem?

Xavier

**Comment by XavierDupessey** [comment]
Same problem with Adal-angular and AdalJs v1.0.13, message:

> AADSTS50058: A silent sign-in request was sent but no user is signed in. The cookies used to represent the user's session were not sent in the request to Azure AD. This can happen if the user is using Internet Explorer or Edge, and the web app sending the silent sign-in request is in different IE security zone than the Azure AD endpoint (login.microsoftonline.com).
> Trace ID: 9391e410-8535-4d8f-abc0-4a2816253a24
> Correlation ID: 8e598faa-60d2-4b57-90d8-57b634141c79
> Timestamp: 2016-12-07 09:59:32Z|login_required|undefined

**Comment by rohitnarula7176** [comment]
@XavierDupessey When you close the browser, the browsing session ends, which results in the cookie related to your login session with AAD getting deleted. If you reopen the browser within one hour, everything will work as expected as when you call acquireToken, adal will attach the valid (not expired ) token from the cache to the outgoing api request without renewing the token by sending the request to AAD. 
But if you select localStorage and reopen the browser after one hour, cache persists and you still have a token in the cache but it is not valid (expired). So when you call acquireToken, adal will send a request to AAD using a hidden iframe to renew the token but it will fail as the cookie got deleted when you closed the browser. This is expected behavior. You can do the following things to resolve your issue:
1.  When you log in for the first time, check the "Keep me signed in" checkbox which will make sure that the cookie persists when you close and reopen the browser.
2.  On your app reload, you can call acquireToken(resource, callback) for the specified resource and if you receive an error similar to the one mentioned above in the callback, just call the login method in your callback function. If you are using the angular wrapper, we will do this for you as long as the route has requireADLogin set to true.

Please let me know if you have any other questions.

**Comment by XavierDupessey** [comment]
Thank you for the explanation.

> 2. On your app reload, you can call acquireToken(resource, callback) for the specified resource and if you receive an error similar to the one mentioned above in the callback, just call the login method in your callback function. If you are using the angular wrapper, we will do this for you as long as the route has requireADLogin set to true.

As I am using the angular wrapper, I expect to be redirected to the login page in my scenario. Why am I not?

I mean, in my situation:
1. the cookie related to my login session with AAD got deleted
2. the token in the cache is not valid because expired

But the angular wrapper does not redirect me to the login page, it seems strange, isn't it?

I also have to find how to activate this "Keep me signed in" option on Azure Portal... It could be a temporary workaround... I do not have it by default:

![adal-login-page](https://cloud.githubusercontent.com/assets/1449888/21101645/6884c480-c07a-11e6-848d-6e2a5ae2c7e9.png)

**Comment by tushargupta51** [comment]
@XavierDupessey You will get redirected to login page only when, after reopening the browser, you navigate to a protected route. Adal will not redirect to login if acquireToken call failed because of cookie not present.

**Comment by XavierDupessey** [comment]
Every routes of the app are protected with `'requireADLogin': true` and I am still not redirected to login page if:

1. the cookie related to my login session with AAD got deleted
and
2. the token in the cache is not valid because expired

**Comment by RobARichardson** [comment]
I'm having this exact same problem in Chrome, Edge, & Firefox using v1.0.13.  And all my routes are protected by `'requiredADLogin': true.`

**Comment by sumeetmd** [comment]
Is there a date by when this issue is expected to be fixed?
My team is using AdalJS v1.0.11 and frequently face this issue when user attempts to re-login into the app after token expiry. Token is issued by Azure AAD. The issue happens on Edge while user is able to login without issues using Internet Explorer or Chrome.
AI logs show the exact same error getting logged i.e. "A silent sign-in request was sent but no user is signed in. The cookies used to represent the user's session were not sent in the request to Azure AD. This can happen if the user is using Internet Explorer or Edge, and the web app sending the silent sign-in request is in different IE security zone than the Azure AD endpoint"

**Comment by johnelliott** [comment]
We are also seeing this error using `adal.js` and localStorage for the cache. At first I didn't even see the error because it was URL-encoded. Now I'm confused because I didn't even think cookies were in play for the way we are using the library

**Comment by duanekord** [comment]
Same here.  Seeing the same issue even after 1.0.13.  Still seeing:A silent sign-in request was sent but no user is signed in. The cookies used to represent the user's session were not sent in the request to Azure AD. This can happen if the user is using Internet Explorer or Edge, and the web app sending the silent sign-in request is in different IE security zone than the Azure AD endpoint

**Comment by johnelliott** [comment]
for the record @duanekord, now I'm not seeing the error. I wasn't calling the `handleWindowCallback`. See here: https://github.com/AzureAD/azure-activedirectory-library-for-js/issues/446#issuecomment-262385450

**Comment by duanekord** [comment]
@sumeetmd 
When you and Thomas were looking into this issue did you come across this  #466 ?

**Comment by urkaver** [comment]
The same problem. I'm using angular-wrapper 1.0.13 and it doestn't handle this problem at all.

**Comment by tushargupta51** [comment]
Is there anyone here who runs into this issue with 'sessionStorage' and re-opening the browser?

TL;dr, This is related to the fact that getCachedUser and getUser does not take into account expiration time of the id token.

Details:
I think this issue occurs when using 'localStorage'. The reason user is not getting redirected to login page is because 'adal.idToken' still has a value when browser is reopened and getCachedUser processed that token and return the user based on it since it does not consider expiration of id token. As a result, userInfo.userName is set to a non-null value, acquireToken request is sent in locationChangeHandler which sets the renewActive flag to true, which in turn causes the "loginHandler" call to be skipped in route/state change handlers.

This issue is related to #471. I am closing this one. Let me know if anyone has any questions.

cc: @rohitnarula7176

**Comment by mathilde-pellerin** [comment]
Hi,
I'm using angular-wrapper 1.0.14 and I get this issue with sessionStorage.
The application is loaded in a page, the user authenticate and then let the page open without touching it for more than an hour. When the user come back and refresh the page, I get this error in the console and the user is not redirected to azure authentication page.
The request causing this error does not appear in the network tab of the developer tools.
On the session storage I can see a key `adal.error` with the value `login_required`.

So I would expect angular-wrapper to redirect the app the login page.

Is this issue fixed in version 1.0.15? It isn't listed in the changelog...

**Comment by mathilde-pellerin** [comment]
@tushargupta51 I checked all the issues related to this one (#471 and from there #456 #486 #445) and all are closed. But I still get the problem, even though I'm using sessionStorage.
Should I open a new issue?

**Comment by sonphnt** [comment]
Hi. I still got the same problem. Somehow, can we make the process of using this library less complicated. For any exceptions like "user is not login" "token time out" "renew token timeout", bla bla. Can we redirect to Microsoft AAD login page where user can sign in to refresh everything again with a simple setting in ADAL init ?

Thanks
